analyzers/src/SonarAnalyzer.VisualBasic/Rules/UseReturnStatement.cs

```diff
@@ -0,0 +1,92 @@
+﻿/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2021 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.VisualBasic;
+using Microsoft.CodeAnalysis.VisualBasic.Syntax;
+using SonarAnalyzer.Common;
+using SonarAnalyzer.Helpers;
+
+namespace SonarAnalyzer.Rules.VisualBasic
+{
+    [DiagnosticAnalyzer(LanguageNames.VisualBasic)]
+    [Rule(DiagnosticId)]
+    public sealed class UseReturnStatement : SonarDiagnosticAnalyzer
+    {
+        private const string DiagnosticId = "S5944";
+        private const string MessageFormat = "Use a 'Return' statement; assigning returned values to function names is obsolete.";
+        private static readonly DiagnosticDescriptor Rule = DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, RspecStrings.ResourceManager);
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+        protected override void Initialize(SonarAnalysisContext context) =>
+            context.RegisterSyntaxNodeActionInNonGenerated(c =>
+            {
+                var method = (MethodStatementSyntax)((MethodBlockSyntax)c.Node).BlockStatement;
+                new Identifierswalker(c, method.Identifier.ValueText).SafeVisit(c.Node);
+            },
+            SyntaxKind.FunctionBlock);
+
+        private class Identifierswalker : VisualBasicSyntaxWalker
+        {
+            private SyntaxNodeAnalysisContext Context { get; }
+            private string Name { get; }
+
+            public Identifierswalker(SyntaxNodeAnalysisContext context, string name)
+            {
+                Context = context;
+                Name = name;
+            }
+
+            public override void VisitIdentifierName(IdentifierNameSyntax node)
+            {
+                if (IsImplicitReturnValue(node))
+                {
+                    if (NoAssignment(node))
```

Comment:
Let's embed this one. It's short and called only once.
```suggestion
                    if (node.Parent is ReturnStatementSyntax || node.Parent is EqualsValueSyntax)
```

Reply:
I strongly disagree; this method explains exactly what is required. The expression it self does not.

Reply:
Let's work on the name and implementation then. Thinking of these testcases, it will produce a wrong message, right? (I'm not able to build your fork for some reason to verify) And there might be more of them.
```
value = ReadReturnValueOnly 
DoSomething(ReadReturnValueOnly)   
```
Let's change the method to `IsAssigmentTarget` and test `node.Parent is AssignmentStatementSyntax assignment && assignment.Left == node`.

Reply:
Lol, I missed your last remark, but draw the same conclusion nevertheless.